File: AmneziaAndroid/core/src/main/java/org/amnezia/core/parsers/AmneziaUrlDecoder.kt
package org.amnezia.core.parsers

import android.util.Base64
import org.amnezia.core.models.ProtocolType
import org.amnezia.core.models.ServerProfile
import org.json.JSONArray
import org.json.JSONObject

object AmneziaUrlDecoder {

    fun decode(input: String): ServerProfile {
        val trimmed = input.trim()
        var payload = trimmed
        if (payload.lowercase().startsWith("vpn://")) {
            payload = payload.substring(6)
                .replace("\u2013", "-")
                .replace("\u2014", "-")
                .replace(" ", "")
                .replace("\t", "")
                .replace("\r", "")
                .replace("\n", "")
        }

        // Try decoding Base64URL
        val data = decodeBase64URL(payload)
        if (data == null) {
            // Check if raw JSON
            if (trimmed.startsWith("{")) {
                val json = JSONObject(trimmed)
                return parseAmneziaJson(json)
            }
            // Check if wg-quick format
            if (trimmed.contains("[Interface]") && trimmed.contains("[Peer]")) {
                return WgQuickConfigParser.parse(trimmed)
            }
            throw IllegalArgumentException("Failed to decode base64 or recognize format")
        }

        // Decompress with Qt zlib
        val decompressed = ZlibHelper.decompress(data)
        val stringContent = String(decompressed, Charsets.UTF_8)

        // Try JSON first (standard Amnezia configs are JSON)
        val jsonObject = try {
            JSONObject(stringContent)
        } catch (e: Exception) {
            null
        }

        if (jsonObject != null) {
            return parseAmneziaJson(jsonObject)
        }

        if (stringContent.contains("[Interface]") && stringContent.contains("[Peer]")) {
            return WgQuickConfigParser.parse(stringContent)
        }

        throw IllegalArgumentException("Unrecognized configuration format")
    }

    fun encodeToUrl(profile: ServerProfile): String {
        val json = profileToAmneziaJson(profile)
        val jsonBytes = json.toString().toByteArray(Charsets.UTF_8)
        val compressed = ZlibHelper.compressQt(jsonBytes)
        val base64Url = encodeBase64URL(compressed)
        return "vpn://$base64Url"
    }

    private fun parseAmneziaJson(root: JSONObject): ServerProfile {
        val description = root.optString("description", "").trim()
        val rootHost = root.optString("hostName", "")

        val dnsList = mutableListOf<String>()
        val dns1 = root.optString("dns1", "")
        val dns2 = root.optString("dns2", "")
        if (dns1.isNotEmpty()) dnsList.add(dns1)
        if (dns2.isNotEmpty()) dnsList.add(dns2)

        val containers = root.optJSONArray("containers")
        if (containers == null || containers.length() == 0) {
            if (root.has("last_config")) {
                return parseLastConfig(root.get("last_config"), root, root, description, rootHost, dnsList)
            }
            throw IllegalArgumentException("Missing containers in Amnezia JSON")
        }

        var targetContainer: JSONObject? = null
        val defaultContainer = root.optString("defaultContainer", "")
        for (i in 0 until containers.length()) {
            val c = containers.getJSONObject(i)
            val name = c.optString("container", "")
            if (name == defaultContainer) {
                targetContainer = c
                break
            }
        }
        if (targetContainer == null) {
            for (i in 0 until containers.length()) {
                val c = containers.getJSONObject(i)
                val name = c.optString("container", "").lowercase()
                if (name.contains("awg") || name.contains("wireguard")) {
                    targetContainer = c
                    break
                }
            }
        }
        if (targetContainer == null) {
            targetContainer = containers.getJSONObject(0)
        }

        val containerType = targetContainer.optString("container", "amnezia-awg")
        val protocolKey = containerType.replace("amnezia-", "")

        var protocolConfig: JSONObject? = targetContainer.optJSONObject(protocolKey)
        if (protocolConfig == null) protocolConfig = targetContainer.optJSONObject("awg")
        if (protocolConfig == null) protocolConfig = targetContainer.optJSONObject("wireguard")
        if (protocolConfig == null) {
            val keys = targetContainer.keys()
            while (keys.hasNext()) {
                val k = keys.next()
                val sub = targetContainer.optJSONObject(k)
                if (sub != null && sub.has("last_config")) {
                    protocolConfig = sub
                    break
                }
            }
        }

        val lastConfigObj = protocolConfig?.opt("last_config")
            ?: throw IllegalArgumentException("Missing last_config in container $containerType")

        return parseLastConfig(lastConfigObj, protocolConfig, root, description, rootHost, dnsList)
    }

    private fun parseLastConfig(
        lastConfigObj: Any,
        containerParams: JSONObject,
        root: JSONObject,
        description: String,
        rootHost: String,
        dnsList: List<String>
    ): ServerProfile {
        val configDict: JSONObject = when (lastConfigObj) {
            is String -> {
                val json = try {
                    JSONObject(lastConfigObj)
                } catch (e: Exception) {
                    null
                }
                if (json != null) {
                    json
                } else if (lastConfigObj.contains("[Interface]") && lastConfigObj.contains("[Peer]")) {
                    return WgQuickConfigParser.parse(lastConfigObj)
                } else {
                    throw IllegalArgumentException("Malformed last_config string")
                }
            }
            is JSONObject -> lastConfigObj
            else -> throw IllegalArgumentException("Malformed last_config")
        }

        fun findString(vararg keys: String): String? {
            for (k in keys) {
                if (configDict.has(k)) {
                    val v = configDict.optString(k, "").trim()
                    if (v.isNotEmpty()) return v
                }
            }
            return null
        }

        val clientPrivKey = findString("clientPrivKey", "client_priv_key", "client_private_key", "PrivateKey")
            ?: throw IllegalArgumentException("Missing client private key")
        val serverPubKey = findString("serverPubKey", "server_pub_key", "server_public_key", "PublicKey")
            ?: throw IllegalArgumentException("Missing server public key")

        var hostName = findString("hostName", "host_name", "server_ip", "serverIp", "Endpoint")
            ?: rootHost.ifEmpty { "127.0.0.1" }
        var port = 51820

        if (configDict.has("port")) {
            port = configDict.optInt("port", 51820)
        } else if (configDict.has("server_port")) {
            port = configDict.optInt("server_port", 51820)
        }

        if (hostName.contains(":") && !hostName.contains("[")) {
            val parts = hostName.split(":")
            if (parts.size == 2) {
                hostName = parts[0]
                parts[1].toIntOrNull()?.let { port = it }
            }
        }

        val rawIp = findString("clientIp", "client_ip", "Address") ?: "10.0.0.2/32"
        val clientIp = if (!rawIp.contains("/")) {
            if (rawIp.contains(":")) "$rawIp/128" else "$rawIp/32"
        } else rawIp
        val psk = findString("pskKey", "psk_key", "preshared_key", "PresharedKey")
        val mtu = configDict.optInt("mtu", 1280)

        // Read AWG obfuscation params from both containerParams and configDict
        fun getAwgInt(key: String): Int? {
            return if (configDict.has(key)) configDict.optInt(key)
            else if (containerParams.has(key)) containerParams.optInt(key)
            else null
        }

        fun getAwgLong(key: String): Long? {
            return if (configDict.has(key)) configDict.optLong(key)
            else if (containerParams.has(key)) containerParams.optLong(key)
            else null
        }

        val jc = getAwgInt("Jc") ?: getAwgInt("jc")
        val jmin = getAwgInt("Jmin") ?: getAwgInt("jmin")
        val jmax = getAwgInt("Jmax") ?: getAwgInt("jmax")
        val s1 = getAwgInt("S1") ?: getAwgInt("s1")
        val s2 = getAwgInt("S2") ?: getAwgInt("s2")
        val h1 = getAwgLong("H1") ?: getAwgLong("h1")
        val h2 = getAwgLong("H2") ?: getAwgLong("h2")
        val h3 = getAwgLong("H3") ?: getAwgLong("h3")
        val h4 = getAwgLong("H4") ?: getAwgLong("h4")

        val isAwg = jc != null || jmin != null || jmax != null || s1 != null || s2 != null ||
                h1 != null || h2 != null || h3 != null || h4 != null
        val protocolType = if (isAwg) ProtocolType.AMNEZIA_WG else ProtocolType.WIREGUARD
        val serverName = if (description.isNotEmpty()) description else "Amnezia ($hostName)"

        return ServerProfile(
            name = serverName,
            protocolType = protocolType,
            host = hostName,
            port = port,
            clientIp = clientIp,
            dnsServers = dnsList.ifEmpty { listOf("1.1.1.1", "1.0.0.1") },
            publicKey = serverPubKey,
            privateKey = clientPrivKey,
            presharedKey = psk,
            allowedIps = listOf("0.0.0.0/0", "::/0"),
            jc = jc,
            jmin = jmin,
            jmax = jmax,
            s1 = s1,
            s2 = s2,
            h1 = h1,
            h2 = h2,
            h3 = h3,
            h4 = h4,
            mtu = mtu
        )
    }

    private fun profileToAmneziaJson(profile: ServerProfile): JSONObject {
        val root = JSONObject()
        root.put("description", profile.name)
        root.put("hostName", profile.host)
        if (profile.dnsServers.isNotEmpty()) {
            root.put("dns1", profile.dnsServers[0])
            if (profile.dnsServers.size > 1) root.put("dns2", profile.dnsServers[1])
        }

        val isAwg = profile.protocolType == ProtocolType.AMNEZIA_WG
        val containerName = if (isAwg) "amnezia-awg" else "amnezia-wireguard"
        root.put("defaultContainer", containerName)

        val lastConfig = JSONObject()
        lastConfig.put("clientPrivKey", profile.privateKey)
        lastConfig.put("serverPubKey", profile.publicKey)
        lastConfig.put("clientIp", profile.clientIp)
        lastConfig.put("port", profile.port)
        lastConfig.put("hostName", profile.host)
        lastConfig.put("mtu", profile.mtu)
        profile.presharedKey?.let { lastConfig.put("pskKey", it) }

        if (isAwg) {
            profile.jc?.let { lastConfig.put("Jc", it) }
            profile.jmin?.let { lastConfig.put("Jmin", it) }
            profile.jmax?.let { lastConfig.put("Jmax", it) }
            profile.s1?.let { lastConfig.put("S1", it) }
            profile.s2?.let { lastConfig.put("S2", it) }
            profile.h1?.let { lastConfig.put("H1", it) }
            profile.h2?.let { lastConfig.put("H2", it) }
            profile.h3?.let { lastConfig.put("H3", it) }
            profile.h4?.let { lastConfig.put("H4", it) }
        }

        val protoDict = JSONObject()
        protoDict.put("last_config", lastConfig)

        val containerObj = JSONObject()
        containerObj.put("container", containerName)
        containerObj.put(if (isAwg) "awg" else "wireguard", protoDict)

        val containers = JSONArray()
        containers.put(containerObj)
        root.put("containers", containers)

        return root
    }

    private fun decodeBase64URL(input: String): ByteArray? {
        var base64 = input
            .replace("-", "+")
            .replace("_", "/")
        val padLength = (4 - (base64.length % 4)) % 4
        if (padLength in 1..3) {
            base64 += "=".repeat(padLength)
        }

        return try {
            java.util.Base64.getDecoder().decode(base64)
        } catch (e: Throwable) {
            try {
                android.util.Base64.decode(base64, android.util.Base64.DEFAULT)
            } catch (e2: Throwable) {
                null
            }
        }
    }

    private fun encodeBase64URL(bytes: ByteArray): String {
        return try {
            java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(bytes).trim()
        } catch (e: Throwable) {
            android.util.Base64.encodeToString(bytes, android.util.Base64.NO_WRAP or android.util.Base64.URL_SAFE)
                .replace("=", "")
                .trim()
        }
    }
}

File: AmneziaCore/Parsers/AmneziaUrlDecoder.swift
import Foundation

public enum AmneziaDecoderError: Error, LocalizedError {
    case invalidUrlScheme
    case base64DecodingFailed
    case jsonParsingFailed(String)
    case unsupportedContainer(String)
    case missingRequiredKeys(String)

    public var errorDescription: String? {
        switch self {
        case .invalidUrlScheme:
            return "URL does not have a valid vpn:// scheme"
        case .base64DecodingFailed:
            return "Failed to decode base64 data"
        case .jsonParsingFailed(let msg):
            return "JSON parsing failed: \(msg)"
        case .unsupportedContainer(let c):
            return "Container '\(c)' is not supported yet"
        case .missingRequiredKeys(let k):
            return "Configuration is missing required keys: \(k)"
        }
    }
}

public final class AmneziaUrlDecoder: Sendable {
    
    /// Decodes a `vpn://...` URL or raw Base64/JSON string into a `ServerProfile`
    public static func decode(_ input: String) throws -> ServerProfile {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        
        // Strip vpn:// prefix if present
        var payload = trimmed
        if payload.lowercased().hasPrefix("vpn://") {
            payload = String(payload.dropFirst(6))
            // Clean up any smart dashes or internal whitespace from iOS pasteboard/texteditor
            payload = payload
                .replacingOccurrences(of: "\u{2013}", with: "-") // En-dash
                .replacingOccurrences(of: "\u{2014}", with: "-") // Em-dash
                .replacingOccurrences(of: " ", with: "")
                .replacingOccurrences(of: "\t", with: "")
                .replacingOccurrences(of: "\r", with: "")
                .replacingOccurrences(of: "\n", with: "")
        }
        
        // Try decoding as Base64URL
        guard let data = decodeBase64URL(payload) else {
            // Check if input is already raw JSON
            if let rawData = trimmed.data(using: .utf8),
               let json = try? JSONSerialization.jsonObject(with: rawData) as? [String: Any] {
                return try parseAmneziaJson(json, rawConfig: trimmed)
            }
            // Check if input is wg-quick format
            if trimmed.contains("[Interface]") && trimmed.contains("[Peer]") {
                return try WgQuickConfigParser.parse(trimmed)
            }
            throw AmneziaDecoderError.base64DecodingFailed
        }
        
        // Try decompressing with Qt qCompress; fallback to raw data
        let jsonData: Data
        if let decompressed = try? ZlibHelper.decompressQt(data) {
            jsonData = decompressed
        } else {
            jsonData = data
        }
        
        guard let jsonObject = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else {
            if let stringContent = String(data: jsonData, encoding: .utf8) {
                // If it decompressed to a wg-quick config string instead of JSON
                if stringContent.contains("[Interface]") && stringContent.contains("[Peer]") {
                    return try WgQuickConfigParser.parse(stringContent)
                }
                throw AmneziaDecoderError.jsonParsingFailed(stringContent)
            }
            throw AmneziaDecoderError.jsonParsingFailed("Invalid UTF-8 data")
        }
        
        return try parseAmneziaJson(jsonObject, rawConfig: trimmed)
    }

    /// Generates an Amnezia `vpn://...` URL from a `ServerProfile`
    public static func encodeToUrl(_ profile: ServerProfile) throws -> String {
        let json = profileToAmneziaJson(profile)
        let jsonData = try JSONSerialization.data(withJSONObject: json, options: [])
        let compressed = try ZlibHelper.compressQt(jsonData)
        let base64Url = encodeBase64URL(compressed)
        return "vpn://\(base64Url)"
    }

    // MARK: - Internal JSON Parsing

    private static func parseAmneziaJson(_ root: [String: Any], rawConfig: String) throws -> ServerProfile {
        let description = (root["description"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        let rootHostName = root["hostName"] as? String
        
        var dnsList = [String]()
        if let dns1 = root["dns1"] as? String, !dns1.isEmpty { dnsList.append(dns1) }
        if let dns2 = root["dns2"] as? String, !dns2.isEmpty { dnsList.append(dns2) }

        // Find primary container
        guard let containers = root["containers"] as? [[String: Any]], !containers.isEmpty else {
            // Check if root itself contains last_config or protocol keys directly
            if let lastConfig = root["last_config"] {
                return try parseLastConfig(lastConfig, containerParams: root, root: root, description: description, rootHost: rootHostName, dnsList: dnsList, rawConfig: rawConfig)
            }
            throw AmneziaDecoderError.missingRequiredKeys("containers")
        }

        let defaultContainer = root["defaultContainer"] as? String
        let targetContainer: [String: Any]
        if let defName = defaultContainer, let match = containers.first(where: { ($0["container"] as? String) == defName }) {
            targetContainer = match
        } else if let match = containers.first(where: {
            let c = ($0["container"] as? String)?.lowercased() ?? ""
            return c.contains("awg") || c.contains("wireguard")
        }) {
            targetContainer = match
        } else {
            targetContainer = containers[0]
        }

        let containerType = (targetContainer["container"] as? String) ?? "amnezia-awg"
        
        // Extract protocol dict (e.g. "awg", "wireguard", etc.)
        var protocolConfig: [String: Any]? = nil
        let protocolKey = containerType.replacingOccurrences(of: "amnezia-", with: "")
        
        if let cfg = targetContainer[protocolKey] as? [String: Any] {
            protocolConfig = cfg
        } else if let cfg = targetContainer["awg"] as? [String: Any] {
            protocolConfig = cfg
        } else if let cfg = targetContainer["wireguard"] as? [String: Any] {
            protocolConfig = cfg
        } else {
            for (_, val) in targetContainer {
                if let subDict = val as? [String: Any], subDict["last_config"] != nil {
                    protocolConfig = subDict
                    break
                }
            }
        }

        guard let protoDict = protocolConfig, let lastConfigObj = protoDict["last_config"] else {
            throw AmneziaDecoderError.missingRequiredKeys("last_config in container \(containerType)")
        }

        return try parseLastConfig(lastConfigObj, containerParams: protoDict, containerType: containerType, root: root, description: description, rootHost: rootHostName, dnsList: dnsList, rawConfig: rawConfig)
    }

    private static func parseLastConfig(
        _ lastConfigObj: Any,
        containerParams: [String: Any] = [:],
        containerType: String = "amnezia-awg",
        root: [String: Any],
        description: String?,
        rootHost: String?,
        dnsList: [String],
        rawConfig: String
    ) throws -> ServerProfile {
        var configDict: [String: Any]
        if let str = lastConfigObj as? String {
            guard let data = str.data(using: .utf8),
                  let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                // If last_config is directly a wg-quick config string (e.g. [Interface]...)
                if str.contains("[Interface]") && str.contains("[Peer]") {
                    return try WgQuickConfigParser.parse(str)
                }
                throw AmneziaDecoderError.jsonParsingFailed("Malformed last_config string")
            }
            configDict = dict
        } else if let dict = lastConfigObj as? [String: Any] {
            configDict = dict
        } else {
            throw AmneziaDecoderError.jsonParsingFailed("last_config is neither string nor dict")
        }

        // Helper to find string across multiple candidate keys (camelCase, snake_case, standard)
        func findString(_ keys: [String]) -> String? {
            for k in keys {
                if let s = configDict[k] as? String, !s.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    return s.trimmingCharacters(in: .whitespacesAndNewlines)
                }
            }
            return nil
        }

        // If configDict has a full wg-quick "config" string and keys are missing, we can also extract from it
        if let embeddedConfig = configDict["config"] as? String,
           (findString(["clientPrivKey", "client_priv_key"]) == nil || findString(["serverPubKey", "server_pub_key"]) == nil),
           let parsed = try? WgQuickConfigParser.parse(embeddedConfig) {
            return parsed
        }

        guard let clientPrivKey = findString(["clientPrivKey", "client_priv_key", "client_private_key", "PrivateKey"]),
              let serverPubKey = findString(["serverPubKey", "server_pub_key", "server_public_key", "PublicKey"]) else {
            throw AmneziaDecoderError.missingRequiredKeys("clientPrivKey, serverPubKey")
        }

        var hostName = findString(["hostName", "host_name", "server_ip", "serverIp", "Endpoint"]) ?? rootHost ?? "127.0.0.1"
        var port: UInt16 = 51820

        if let pInt = configDict["port"] as? Int ?? configDict["server_port"] as? Int {
            port = UInt16(pInt)
        } else if let pStr = findString(["port", "server_port"]), let p = UInt16(pStr) {
            port = p
        }

        // If hostName contains port (e.g. 1.2.3.4:51820)
        if hostName.contains(":") && !hostName.contains("[") {
            let parts = hostName.split(separator: ":")
            if parts.count == 2 {
                hostName = String(parts[0])
                if let p = UInt16(parts[1]) {
                    port = p
                }
            }
        }

        var clientAddresses = [String]()
        if let ipStr = findString(["clientIp", "client_ip", "Address"]) {
            var addr = ipStr
            if !addr.contains("/") {
                addr = addr.contains(":") ? "\(addr)/128" : "\(addr)/32"
            }
            clientAddresses.append(addr)
        } else if let ips = (configDict["clientIp"] as? [String]) ?? (configDict["client_ip"] as? [String]) {
            clientAddresses.append(contentsOf: ips.map { $0.contains("/") ? $0 : ($0.contains(":") ? "\($0)/128" : "\($0)/32") })
        }

        var allowedIps = [String]()
        if let aIps = (configDict["allowedIps"] as? [String]) ?? (configDict["allowed_ips"] as? [String]), !aIps.isEmpty {
            allowedIps = aIps
        } else if let aIpsStr = findString(["allowedIps", "allowed_ips", "AllowedIPs"]) {
            allowedIps = aIpsStr.components(separatedBy: ",").map { $0.trimmingCharacters(in: .whitespaces) }
        } else {
            allowedIps = ["0.0.0.0/0", "::/0"]
        }

        let psk = findString(["pskKey", "psk_key", "preshared_key", "PresharedKey"])

        var mtu: UInt16?
        if let mtuInt = configDict["mtu"] as? Int ?? configDict["MTU"] as? Int {
            mtu = UInt16(mtuInt)
        } else if let mtuStr = findString(["mtu", "MTU"]), let m = UInt16(mtuStr) {
            mtu = m
        }

        var keepalive: UInt16?
        if let kInt = configDict["persistentKeepAlive"] as? Int ?? configDict["persistent_keep_alive"] as? Int {
            keepalive = UInt16(kInt)
        } else if let kStr = findString(["persistentKeepAlive", "persistent_keep_alive", "PersistentKeepalive"]), let k = UInt16(kStr) {
            keepalive = k
        }

        // Parse AWG obfuscation parameters from both container parameters and last_config
        var mergedAwgDict = containerParams
        for (k, v) in configDict {
            mergedAwgDict[k] = v
        }
        let awg = parseAwgParams(from: mergedAwgDict)
        let isAwg = containerType.contains("awg") || awg.hasObfuscation
        let protocolType: ProtocolType = isAwg ? .amneziaWg : .wireGuard
        let serverName = description?.isEmpty == false ? description! : "Amnezia (\(hostName))"

        return ServerProfile(
            name: serverName,
            protocolType: protocolType,
            endpointHost: hostName,
            endpointPort: port,
            clientPrivateKey: clientPrivKey,
            serverPublicKey: serverPubKey,
            presharedKey: psk,
            clientAddresses: clientAddresses.isEmpty ? ["10.8.0.2/32"] : clientAddresses,
            dnsServers: dnsList.isEmpty ? ["1.1.1.1", "1.0.0.1"] : dnsList,
            allowedIps: allowedIps,
            mtu: mtu,
            persistentKeepalive: keepalive,
            awgParameters: awg.hasObfuscation ? awg : nil,
            rawConfig: rawConfig
        )
    }

    private static func parseAwgParams(from dict: [String: Any]) -> AwgParameters {
        func getUInt16(_ key: String) -> UInt16? {
            if let val = dict[key] as? Int { return UInt16(val) }
            if let val = dict[key.lowercased()] as? Int { return UInt16(val) }
            if let str = dict[key] as? String, let val = UInt16(str.trimmingCharacters(in: .whitespaces)) { return val }
            if let str = dict[key.lowercased()] as? String, let val = UInt16(str.trimmingCharacters(in: .whitespaces)) { return val }
            return nil
        }
        func getString(_ key: String) -> String? {
            if let str = dict[key] as? String, !str.trimmingCharacters(in: .whitespaces).isEmpty {
                return str.trimmingCharacters(in: .whitespaces)
            }
            if let str = dict[key.lowercased()] as? String, !str.trimmingCharacters(in: .whitespaces).isEmpty {
                return str.trimmingCharacters(in: .whitespaces)
            }
            if let num = dict[key] as? Int { return "\(num)" }
            return nil
        }

        return AwgParameters(
            jc: getUInt16("Jc"),
            jmin: getUInt16("Jmin"),
            jmax: getUInt16("Jmax"),
            s1: getUInt16("S1"),
            s2: getUInt16("S2"),
            s3: getUInt16("S3"),
            s4: getUInt16("S4"),
            h1: getString("H1"),
            h2: getString("H2"),
            h3: getString("H3"),
            h4: getString("H4"),
            i1: getString("I1"),
            i2: getString("I2"),
            i3: getString("I3"),
            i4: getString("I4"),
            i5: getString("I5")
        )
    }

    private static func profileToAmneziaJson(_ profile: ServerProfile) -> [String: Any] {
        var lastConfigDict: [String: Any] = [
            "clientPrivKey": profile.clientPrivateKey,
            "serverPubKey": profile.serverPublicKey,
            "clientIp": profile.clientAddresses.first ?? "10.8.0.2/32",
            "hostName": profile.endpointHost,
            "port": Int(profile.endpointPort),
            "allowedIps": profile.allowedIps
        ]
        
        if let psk = profile.presharedKey { lastConfigDict["pskKey"] = psk }
        if let mtu = profile.mtu { lastConfigDict["mtu"] = Int(mtu) }
        if let keepalive = profile.persistentKeepalive { lastConfigDict["persistentKeepAlive"] = Int(keepalive) }
        
        if let awg = profile.awgParameters {
            for (k, v) in awg.configDictionary {
                lastConfigDict[k] = v
            }
        }
        
        let lastConfigJsonData = (try? JSONSerialization.data(withJSONObject: lastConfigDict, options: [])) ?? Data()
        let lastConfigJsonString = String(data: lastConfigJsonData, encoding: .utf8) ?? "{}"
        
        let containerName = (profile.protocolType == .amneziaWg) ? "amnezia-awg" : "amnezia-wireguard"
        let protocolKey = (profile.protocolType == .amneziaWg) ? "awg" : "wireguard"
        
        let container: [String: Any] = [
            "container": containerName,
            protocolKey: [
                "last_config": lastConfigJsonString
            ]
        ]
        
        var root: [String: Any] = [
            "containers": [container],
            "defaultContainer": containerName,
            "description": profile.name,
            "hostName": profile.endpointHost
        ]
        
        if profile.dnsServers.indices.contains(0) { root["dns1"] = profile.dnsServers[0] }
        if profile.dnsServers.indices.contains(1) { root["dns2"] = profile.dnsServers[1] }
        
        return root
    }

    // MARK: - Base64URL Helpers

    public static func decodeBase64URL(_ input: String) -> Data? {
        var base64 = input
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = base64.count % 4
        if remainder > 0 {
            base64.append(String(repeating: "=", count: 4 - remainder))
        }
        return Data(base64Encoded: base64)
    }

    public static func encodeBase64URL(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
